File: editor/src/Packaging/PackageLayout.cs
// ============================================================
//  PackageLayout.cs — 配布パッケージのフォルダ構成【正典（エディタ側）】
//
//  【役割】
//  パッケージ出力フォルダ（`{出力先}/{ゲーム名}/`）の中に作るサブフォルダの
//  名前と、そこへのパスを組み立てる純関数だけを持つ。
//  「何をコピーするか」は各パッケージャ（ScriptPackager / DotnetRuntimeBundler）の
//  責務で、ここは置き場の知識だけを担当する。
//
//  【配布物の構成】
//  ```text
//  {ゲーム名}/
//    {ゲーム名}.exe          … 実行ファイル
//    assets.pak              … アセット（PAK）
//    bin/                    … 実行に必要な副次ファイル（パッケージ化で同梱する）
//      SEEDScripting.dll / SEEDScripting.runtimeconfig.json / SEEDScripting.deps.json
//      Microsoft.CodeAnalysis*.dll / SEEDUserScripts.dll
//      dotnet/               … 同梱 .NET ランタイム（self-contained 配布）
//    caches/                 … 実行時生成（モデル派生キャッシュ / pipeline_cache.bin）
//    logs/                   … 実行時生成（起動ログ seed_*.log）
//    saved/                  … 実行時生成（セーブデータ save.json）
//  ```
//
//  【空フォルダを作らない方針】
//  `caches` / `logs` / `saved` は **パッケージ化では作らない**。
//  空フォルダは zip 化・展開で落ちることが多く、「あるはず」を前提にすると
//  配布先でだけ壊れる。必要になった時点でランタイムが作る。
//  同じ理由で、これらは**削除もしない**（利用者のセーブ・ログが入っているため）。
//
//  【ランタイム側の対応物】
//  `runtime/src/engine/core/package_layout.rs` が同じ名前の定数を持つ。
//  どちらかを変えたら必ず両方直すこと（名前がずれると配布物だけが壊れる）。
//
//  【WPF 非依存】
//  PackagingCollectorTests / ScriptPrecompileTests からリンクで取り込んでテストするため、
//  このファイルは WPF 型に依存してはならない（依存した瞬間にテストが壊れて気付ける）。
// ============================================================

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace SEEDEditor.Packaging;

/// <summary>配布パッケージのフォルダ構成（サブフォルダ名とパス組み立て）。</summary>
public static class PackageLayout
{
    // ── フォルダ名（ランタイム側 package_layout.rs と一致必須） ──

    /// <summary>
    /// 実行に必要な副次ファイル（スクリプトホスト DLL・同梱 .NET）を入れるフォルダ名。
    /// ランタイム側 <c>package_layout::BIN_DIR_NAME</c> と一致必須。
    /// </summary>
    public const string BinDirName = "bin";

    /// <summary>
    /// 実行時に生成する派生データキャッシュのフォルダ名。
    /// ランタイム側 <c>package_layout::CACHES_DIR_NAME</c> と一致必須。
    /// </summary>
    public const string CachesDirName = "caches";

    /// <summary>
    /// 起動ログのフォルダ名。
    /// ランタイム側 <c>package_layout::LOGS_DIR_NAME</c> と一致必須。
    /// </summary>
    public const string LogsDirName = "logs";

    /// <summary>
    /// セーブデータのフォルダ名（パッケージ実行のみ）。
    /// ランタイム側 <c>package_layout::SAVED_DIR_NAME</c> と一致必須。
    /// </summary>
    public const string SavedDirName = "saved";

    /// <summary>
    /// 実行時に生成されるフォルダ（パッケージ化では作らない・消さない）。
    /// 中身は利用者のセーブ・ログ・キャッシュなので、後始末の対象から必ず外す。
    /// </summary>
    public static readonly IReadOnlyList<string> RuntimeGeneratedDirNames =
        [CachesDirName, LogsDirName, SavedDirName];

    // ── 旧レイアウトの残骸（出力フォルダ直下に散っていたもの） ──

    /// <summary>同梱 .NET ルートのフォルダ名（旧レイアウトでは出力フォルダ直下にあった）。</summary>
    private const string BundledDotnetDirName = "dotnet";

    /// <summary>旧レイアウトで出力フォルダ直下へコピーしていたファイルの拡張子（DLL）。</summary>
    private const string ManagedAssemblyExtension = ".dll";

    /// <summary>
    /// 旧レイアウトで出力フォルダ直下へコピーしていた設定ファイルの接尾辞。
    ///
    /// <para>
    /// <c>*.json</c> をまとめて対象にはしない。出力フォルダは利用者が中身を足せる
    /// 場所であり、無関係な JSON を巻き込む危険があるため、
    /// スクリプトホストが実際に置く 2 種類だけを名指しする。
    /// </para>
    /// </summary>
    private static readonly string[] LegacyConfigSuffixes =
        [".deps.json", ".runtimeconfig.json"];

    // ── パス組み立て（純関数） ───────────────────────────────

    /// <summary>副次ファイルフォルダ <c>{gameOutDir}/bin</c> を返す【純関数】。</summary>
    /// <param name="gameOutDir">パッケージ出力フォルダ。</param>
    /// <returns><c>{gameOutDir}/bin</c>。</returns>
    public static string BinDirectory(string gameOutDir) =>
        Path.Combine(gameOutDir, BinDirName);

    // ── 旧レイアウトの後始末（判定は純関数） ─────────────────

    /// <summary>
    /// 出力フォルダ直下のファイル名が「旧レイアウトの残骸」かを判定する【純関数】。
    ///
    /// <para>
    /// 新レイアウトでは直下に置くのは実行ファイル（<c>.exe</c>）と <c>assets.pak</c> だけ。
    /// DLL と <c>*.deps.json</c> / <c>*.runtimeconfig.json</c> は必ず <c>bin/</c> 側にあるため、
    /// 直下に残っていれば古いパッケージ化の名残と断定できる。
    /// </para>
    /// </summary>
    /// <param name="fileName">ファイル名（パスではなく名前だけ）。</param>
    /// <returns>削除してよいなら true。</returns>
    public static bool IsLegacyLeftoverFile(string fileName)
    {
        if (string.IsNullOrEmpty(fileName)) return false;

        if (fileName.EndsWith(ManagedAssemblyExtension, StringComparison.OrdinalIgnoreCase))
            return true;

        return LegacyConfigSuffixes.Any(
            suffix => fileName.EndsWith(suffix, StringComparison.OrdinalIgnoreCase));
    }

    /// <summary>
    /// 出力フォルダ直下のフォルダ名が「旧レイアウトの残骸」かを判定する【純関数】。
    ///
    /// <para>
    /// 対象は旧配置の <c>dotnet/</c> だけ。<c>caches</c> / <c>logs</c> / <c>saved</c> は
    /// 利用者データなので絶対に対象にしない。
    /// </para>
    /// </summary>
    /// <param name="dirName">フォルダ名（パスではなく名前だけ）。</param>
    /// <returns>削除してよいなら true。</returns>
    public static bool IsLegacyLeftoverDirectory(string dirName) =>
        string.Equals(dirName, BundledDotnetDirName, StringComparison.OrdinalIgnoreCase);

    /// <summary>
    /// 出力フォルダ直下の一覧から、削除すべき旧レイアウトの残骸を選ぶ【純関数】。
    /// </summary>
    /// <param name="fileNames">直下のファイル名一覧。</param>
    /// <param name="dirNames">直下のフォルダ名一覧。</param>
    /// <returns>削除対象の名前（入力の部分集合。ファイル → フォルダの順）。</returns>
    public static IReadOnlyList<string> SelectLegacyLeftovers(
        IEnumerable<string> fileNames,
        IEnumerable<string> dirNames)
    {
        var result = new List<string>();
        result.AddRange(fileNames.Where(IsLegacyLeftoverFile));
        result.AddRange(dirNames.Where(IsLegacyLeftoverDirectory));
        return result;
    }

    /// <summary>
    /// 出力フォルダ直下から旧レイアウトの残骸を削除する。
    ///
    /// <para>
    /// 配置を <c>bin/</c> へ移したあと、同じフォルダへ再度パッケージ化すると
    /// 古い DLL が直下に残り続ける。ランタイムは <c>bin/</c> しか見ないので
    /// 実行には影響しないが、「exe の隣に DLL が散らかる」状態が消えないうえ、
    /// 利用者から見て新旧どちらが本物か分からなくなるため必ず消す。
    /// </para>
    /// <para>
    /// 削除の失敗（他プロセスが開いている等）は握りつぶさずログに出すが、
    /// パッケージ化そのものは止めない。
    /// </para>
    /// </summary>
    /// <param name="gameOutDir">パッケージ出力フォルダ。</param>
    /// <param name="log">ログ出力（UI へ 1 行ずつ流す）。</param>
    /// <returns>削除できた件数。</returns>
    public static int RemoveLegacyLayout(string gameOutDir, Action<string> log)
    {
        if (!Directory.Exists(gameOutDir)) return 0;

        var fileNames = Directory.EnumerateFiles(gameOutDir).Select(Path.GetFileName)!;
        var dirNames  = Directory.EnumerateDirectories(gameOutDir).Select(Path.GetFileName)!;

        var targets = SelectLegacyLeftovers(fileNames!, dirNames!);
        if (targets.Count == 0) return 0;

        log($"旧レイアウトの残骸を削除します（{targets.Count} 件）");

        int removed = 0;
        foreach (var name in targets)
        {
            var path = Path.Combine(gameOutDir, name);
            try
            {
                if (Directory.Exists(path)) Directory.Delete(path, recursive: true);
                else                        File.Delete(path);

                removed++;
                log($"    削除: {name}");
            }
            catch (Exception ex)
            {
                log($"    ⚠ 削除できませんでした: {name}（{ex.Message}）");
            }
        }

        return removed;
    }
}

File: runtime/src/engine/core/package_layout.rs
// ============================================================
//  package_layout.rs — 配布パッケージのフォルダ構成【正典】
//
//  【役割】
//  配布物（{ゲーム名}/ フォルダ）の中に置くサブフォルダの名前と、
//  実行ファイルの位置からそれぞれの絶対パスを求める純関数だけを持つ。
//  ファイル I/O は一切行わない（作る・消すは各機能の責務）。
//
//  【なぜ 1 モジュールに集めるのか】
//  以前は「スクリプトホスト DLL は exe の隣」「セーブは exe の隣の save/」
//  「モデルキャッシュは assets の隣の cache/」と、置き場の知識が
//  機能ごとに散らばっていた。その結果 exe と同じ階層に DLL が数十個
//  散らかり、どれが配布物でどれが実行時生成物かを利用者が判別できなかった。
//  ここを唯一の定義箇所にして、フォルダ名を変えるときの変更点を 1 つにする。
//
//  【配布物の構成】
//  ```text
//  {ゲーム名}/
//    {ゲーム名}.exe          … 実行ファイル
//    assets.pak              … アセット（PAK）
//    bin/                    … 実行に必要な副次ファイル（配布時に同梱する）
//      SEEDScripting.dll / SEEDScripting.runtimeconfig.json / SEEDScripting.deps.json
//      Microsoft.CodeAnalysis*.dll / SEEDUserScripts.dll
//      dotnet/               … 同梱 .NET ランタイム（self-contained 配布）
//    caches/                 … 実行時生成（モデル派生キャッシュ *.smdl / pipeline_cache.bin）
//    logs/                   … 実行時生成（起動ログ seed_*.log）
//    saved/                  … 実行時生成（セーブデータ save.json）
//  ```
//
//  【空フォルダを作らない方針】
//  `caches` / `logs` / `saved` は **パッケージ化では作らない**。
//  空フォルダは zip 化・展開で落ちることが多く、「あるはず」を前提にすると
//  配布先でだけ壊れる。必要になった時点でランタイムが `create_dir_all` する。
//
//  【エディタ側の対応物】
//  `editor/src/Packaging/PackageLayout.cs` が同じ名前の定数を持つ。
//  どちらかを変えたら必ず両方直すこと（名前がずれると配布物だけが壊れる）。
// ============================================================

use std::path::{Path, PathBuf};

// ============================================================
//  フォルダ名（配布物の中に作るサブフォルダ）
// ============================================================

/// 実行に必要な副次ファイル（スクリプトホスト DLL・同梱 .NET）を入れるフォルダ名。
///
/// エディタ側 `PackageLayout.BinDirName` と一致必須。
pub const BIN_DIR_NAME: &str = "bin";

/// 実行時に生成する派生データキャッシュのフォルダ名。
///
/// 中身は「消しても再生成できるもの」だけに限る（消せることが利用者への契約）。
/// エディタ側 `PackageLayout.CachesDirName` と一致必須。
pub const CACHES_DIR_NAME: &str = "caches";

/// 起動ログを入れるフォルダ名。
///
/// 実際に書き出すのは `core::startup_log`（候補の決定は `startup_log::log_path`）。
/// エディタ側 `PackageLayout.LogsDirName` と一致必須。
pub const LOGS_DIR_NAME: &str = "logs";

/// セーブデータを入れるフォルダ名（パッケージ実行のみ）。
///
/// エディタ Play は配布物ではないので別の場所（`runtime/save/`）を使う。
/// 判定は `core::save::path::decide_save_dir`。
/// エディタ側 `PackageLayout.SavedDirName` と一致必須。
pub const SAVED_DIR_NAME: &str = "saved";

// ============================================================
//  パス組み立て（純関数）
// ============================================================

/// 副次ファイルフォルダ `{exe のフォルダ}/bin` を返す【純関数】。
///
/// # 引数
/// * `exe_dir` - 実行ファイルのあるフォルダ
pub fn bin_dir(exe_dir: &Path) -> PathBuf {
    exe_dir.join(BIN_DIR_NAME)
}

/// キャッシュフォルダ `{exe のフォルダ}/caches` を返す【純関数】。
///
/// # 引数
/// * `exe_dir` - 実行ファイルのあるフォルダ
pub fn caches_dir(exe_dir: &Path) -> PathBuf {
    exe_dir.join(CACHES_DIR_NAME)
}

/// ログフォルダ `{exe のフォルダ}/logs` を返す【純関数】。
///
/// # 引数
/// * `exe_dir` - 実行ファイルのあるフォルダ
pub fn logs_dir(exe_dir: &Path) -> PathBuf {
    exe_dir.join(LOGS_DIR_NAME)
}

/// セーブフォルダ `{exe のフォルダ}/saved` を返す【純関数】。
///
/// # 引数
/// * `exe_dir` - 実行ファイルのあるフォルダ
pub fn saved_dir(exe_dir: &Path) -> PathBuf {
    exe_dir.join(SAVED_DIR_NAME)
}

/// 実行時生成キャッシュの置き場を決める【純関数】。
///
/// パッケージ実行かどうかで置き場が変わる、という判断だけをここに集める。
/// 「開発時の置き場」は用途ごとに違う（モデルキャッシュは `{assets}/../cache`、
/// パイプラインキャッシュは exe の隣）ため、呼び出し側が `dev_dir` で渡す。
///
/// # 引数
/// * `packaged` - パッケージ実行か（`asset_fs::is_packaged()`）
/// * `exe_dir`  - 実行ファイルのあるフォルダ（取得できなければ `None`）
/// * `dev_dir`  - パッケージ実行でないときに使う置き場（決められなければ `None`）
///
/// # 戻り値
/// キャッシュを置くフォルダ。決められない場合は `None`
/// （呼び出し側はキャッシュを諦める＝機能は落とさない）。
pub fn decide_cache_dir(
    packaged: bool,
    exe_dir: Option<&Path>,
    dev_dir: Option<PathBuf>,
) -> Option<PathBuf> {
    // パッケージ実行では配布フォルダの中の caches/ に集約する。
    // 実行ファイルの位置が取れないときだけ開発時の置き場へ落ちる
    // （ゲームを起動できなくするほどの事情ではない）。
    if packaged {
        if let Some(dir) = exe_dir {
            return Some(caches_dir(dir));
        }
    }
    dev_dir
}

// ============================================================
//  ユニットテスト（すべて純関数なので完全に検証できる）
// ============================================================

#[cfg(test)]
mod tests {
    use super::*;

    /// 各フォルダが exe の直下に 1 段だけ作られること。
    #[test]
    fn directories_are_direct_children_of_exe_dir() {
        let exe = Path::new("D:/Games/MyGame");
        assert_eq!(bin_dir(exe), Path::new("D:/Games/MyGame/bin"));
        assert_eq!(caches_dir(exe), Path::new("D:/Games/MyGame/caches"));
        assert_eq!(logs_dir(exe), Path::new("D:/Games/MyGame/logs"));
        assert_eq!(saved_dir(exe), Path::new("D:/Games/MyGame/saved"));
    }

    /// フォルダ名がエディタ側（PackageLayout.cs）の規約と一致していること。
    /// 名前がずれるとビルドは通るのに配布物だけが壊れるため、文字列で固定する。
    #[test]
    fn dir_names_match_editor_contract() {
        assert_eq!(BIN_DIR_NAME, "bin");
        assert_eq!(CACHES_DIR_NAME, "caches");
        assert_eq!(LOGS_DIR_NAME, "logs");
        assert_eq!(SAVED_DIR_NAME, "saved");
    }

    /// パッケージ実行では exe の隣の caches/ を使う（開発時の置き場より優先）。
    #[test]
    fn packaged_cache_dir_uses_exe_caches() {
        let dir = decide_cache_dir(
            true,
            Some(Path::new("D:/Games/MyGame")),
            Some(PathBuf::from("C:/repo/runtime/cache")),
        );
        assert_eq!(dir, Some(PathBuf::from("D:/Games/MyGame/caches")));
    }

    /// 非パッケージ実行では開発時の置き場をそのまま使う。
    #[test]
    fn dev_cache_dir_is_passed_through() {
        let dir = decide_cache_dir(
            false,
            Some(Path::new("C:/repo/runtime/target/debug")),
            Some(PathBuf::from("C:/repo/runtime/cache")),
        );
        assert_eq!(dir, Some(PathBuf::from("C:/repo/runtime/cache")));
    }

    /// パッケージ実行でも exe の位置が取れなければ開発時の置き場へ落ちる。
    #[test]
    fn packaged_without_exe_dir_falls_back_to_dev_dir() {
        let dir = decide_cache_dir(true, None, Some(PathBuf::from("C:/fallback/cache")));
        assert_eq!(dir, Some(PathBuf::from("C:/fallback/cache")));
    }

    /// どちらも決められないときは None（キャッシュ無しで動く）。
    #[test]
    fn no_information_yields_none() {
        assert_eq!(decide_cache_dir(true, None, None), None);
        assert_eq!(decide_cache_dir(false, Some(Path::new("C:/game")), None), None);
    }
}
